Mathlib/Tactic/FailIfNoProgress.lean

```diff
@@ -4,63 +4,200 @@ Released under Apache 2.0 license as described in the file LICENSE.
+* Properties of `LocalDecl`s to compare (via extending `LocalDeclComparisonConfig`):
+  * `checkUserName := true`
+  * `checkType := true`
+  * `checkLetValue := true`
+  * `checkBinderInfo := true`
+  * `checkIndex := false`
+  * `checkFVarId := false`
+  * `checkLocalDeclKind := false`
+-/
+structure Config extends ExprComparisonConfig, MVarIdComparisonConfig where
+  mode : FailIfNoProgress.Mode := .normal
+deriving Repr, Inhabited
+
+/-! ## Preset configs -/
+
+/-- Check for any changes whatsoever. -/
+def Config.anyChanges : Config := { MVarIdComparisonConfig.anyChanges with eqKind := .beq }
 
-open Lean Meta Elab Tactic
+/-- Only compare expressions in the target and local context. (Does not include implementation
+details; does check local instances.) -/
+def Config.onlyExprs : Config := { MVarIdComparisonConfig.onlyExprs with }
 
-/-- `fail_if_no_progress tacs` evaluates `tacs`, and fails if no progress is made on the main goal
-or the local context at reducible transparency. -/
-syntax (name := failIfNoProgress) "fail_if_no_progress " tacticSeq : tactic
+/-! ## TacticM implementations -/
 
-/-- `lctxIsDefEq l₁ l₂` compares two lists of `Option LocalDecl`s (as returned from e.g.
-`(← (← getMainGoal).getDecl).lctx.decls.toList`). It returns `true` if they contain expressions of
-the same type in the same order (up to defeq), and `false` otherwise. -/
-def lctxIsDefEq : (l₁ l₂ : List (Option LocalDecl)) → MetaM Bool
-  | some d₁ :: l₁, some d₂ :: l₂ => do
-    unless (← withNewMCtxDepth <| isDefEq d₁.type d₂.type) do
-      return false
-    lctxIsDefEq l₁ l₂
-  | none :: l₁, none :: l₂ => lctxIsDefEq l₁ l₂
-  | [], [] => return true
-  | _, _ => return false
+/-- Run `tacs : TacticM Unit` on `goal`, and fail if no progress is made. Return the resulting list
+of goals otherwise. By default, this compares each of their types and local contexts before and
+after `tacs` is run; if any changes are seen, "progress" has been made, and the tactics succeed.
+Otherwise, we fail.
 
-/-- Run `tacs : TacticM Unit` on `goal`, and fail if no progress is made. -/
-def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit) : TacticM (List MVarId) := do
+`cfg` can be used to specify what kind of comparison to perform when checking for "progress". By
+default, only "observable" changes are checked. For instance, internal `MVarId`s and `FVarId`s are
+not checked, and implementation details are ignored. Expressions are by default compared with
+`isDefEq` at reducible transparency. Any change in the order or number of goals counts as progress.
+
+The config preset `FailIfNoProgress.Config.anyChanges` can be used to detect any change;
+`FailIfNoProgress.Config.onlyExprs` will only compare expressions in the target and local context.
+
+For more information, see the documentation for `FailIfNoProgress.Config`.
+-/
+def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit)
+    (cfg : FailIfNoProgress.Config := {}) : TacticM (List MVarId) := do
```

Comment:
This seems a weird type signature. Perhaps that return type should be `MetaM (List MVarId)`?

Possibly also `tacs` should be a `MVarId \to MetaM (List MVarId)`?

Reply:
Does this make sense? My objections are:
* This is all operating in `MetaM` anyway, using `run`. (maybe TermElabM, whatever)
* TacticM shouldn't ever be returning `List MVarId`: that's what its state is for!
* I want a purely `MetaM` level version of all this anyway.

Reply:
You're right! Looking back I actually don't know why I even defined `runAndFailIfNoProgress`. I replaced it with `MVarId.failIfNoProgress`, `MVarId.failIfNoProgress'`, and `MVarId.failIfNoProgress1` in analogy to the corresponding `liftMetaTactic`s, all in `MetaM`.